core/network/address.go

```diff
@@ -284,6 +284,11 @@ type ProviderAddress struct {
 	MachineAddress
 	SpaceName       SpaceName
 	ProviderSpaceID Id
+
+	// The spaceID, if known, can be optionally stored when constructing
+	// the address to avoid additional lookups when converting provider
+	// addresses into SpaceAddresses.
+	spaceID string
```

Comment:
I really don't want to add this here.

Consider something like this:
```diff
diff --git a/apiserver/facades/controller/instancepoller/instancepoller.go b/apiserver/facades/controller/instancepoller/instancepoller.go
index f1d3c1392d..9a62e42636 100644
--- a/apiserver/facades/controller/instancepoller/instancepoller.go
+++ b/apiserver/facades/controller/instancepoller/instancepoller.go
@@ -158,10 +158,7 @@ func (a *InstancePollerAPI) SetProviderNetworkConfig(req params.SetProviderNetwo
 			continue
 		}
 
-		// spaceIDs have already been resolved for us by the above
-		// function call; we can safely provide a nil lookup to the
-		// converter and ignore errors.
-		newSpaceAddrs, _ := newProviderAddrs.ToSpaceAddresses(nil)
+		newSpaceAddrs, _ := newProviderAddrs.ToSpaceAddresses(spaceInfos)
 
 		modified, err := maybeUpdateMachineProviderAddresses(machine, newSpaceAddrs)
 		if err != nil {
@@ -300,7 +297,7 @@ func mapNetworkConfigsToProviderAddresses(cfgs []params.NetworkConfig, spaceInfo
 
 			addrs = append(
 				addrs,
-				network.NewScopedProviderAddressFromSpaceInfo(addr.Value, addr.Scope, spaceInfo),
+				network.NewScopedProviderAddressInSpace(addr.Value, addr.Scope, spaceInfo),
 			)
 		}
 
@@ -320,7 +317,7 @@ func mapNetworkConfigsToProviderAddresses(cfgs []params.NetworkConfig, spaceInfo
 			}
 			addrs = append(
 				addrs,
-				network.NewScopedProviderAddressFromSpaceInfo(addr.Value, addr.Scope, spaceInfo),
+				network.NewScopedProviderAddressInSpace(addr.Value, addr.Scope, spaceInfo),
 			)
 		}
 	}
diff --git a/core/network/address.go b/core/network/address.go
index d5a3b9da16..b3de64374a 100644
--- a/core/network/address.go
+++ b/core/network/address.go
@@ -284,11 +284,6 @@ type ProviderAddress struct {
 	MachineAddress
 	SpaceName       SpaceName
 	ProviderSpaceID Id
-
-	// The spaceID, if known, can be optionally stored when constructing
-	// the address to avoid additional lookups when converting provider
-	// addresses into SpaceAddresses.
-	spaceID string
 }
 
 // GoString implements fmt.GoStringer.
@@ -348,15 +343,12 @@ func NewProviderAddressInSpace(spaceName string, value string) ProviderAddress {
 	}
 }
 
-// NewScopedProviderAddressFromSpaceInfo creates a new SpaceAddress by
-// encapsulating a MachineAddress and populating its space name, juju space ID
-// and provider space ID values from the provided SpaceInfo instance.
-func NewScopedProviderAddressFromSpaceInfo(value string, scope Scope, spaceInfo *SpaceInfo) ProviderAddress {
+// NewScopedProviderAddressInSpace...
+func NewScopedProviderAddressInSpace(value string, scope Scope, spaceInfo *SpaceInfo) ProviderAddress {
 	return ProviderAddress{
 		MachineAddress:  NewScopedMachineAddress(value, scope),
 		SpaceName:       spaceInfo.Name,
 		ProviderSpaceID: spaceInfo.ProviderId,
-		spaceID:         spaceInfo.ID,
 	}
 }
 
@@ -405,38 +397,20 @@ func (pas ProviderAddresses) ToSpaceAddresses(lookup SpaceLookup) (SpaceAddresse
 		return nil, nil
 	}
 
-	// If we have already resolved the space IDs we can do away with the
-	// extra lookup.
-	var requireNameLookups bool
-	for _, pa := range pas {
-		if pa.spaceID == "" {
-			requireNameLookups = true
-			break
-		}
+	spaceInfos, err := lookup.AllSpaceInfos()
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	var idFor map[SpaceName]string
-	if requireNameLookups {
-		spaceInfos, err := lookup.AllSpaceInfos()
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-
-		idFor = make(map[SpaceName]string, len(spaceInfos))
-		for _, si := range spaceInfos {
-			idFor[si.Name] = si.ID
-		}
+	idFor := make(map[SpaceName]string, len(spaceInfos))
+	for _, si := range spaceInfos {
+		idFor[si.Name] = si.ID
 	}
 
 	sas := make(SpaceAddresses, len(pas))
 	for i, pa := range pas {
 		sas[i] = SpaceAddress{MachineAddress: pa.MachineAddress}
 		if pa.SpaceName != "" {
-			if pa.spaceID != "" {
-				sas[i].SpaceID = pa.spaceID
-				continue
-			}
-
 			id, ok := idFor[pa.SpaceName]
 			if !ok {
 				return nil, errors.NotFoundf("space with name %q", pa.SpaceName)
diff --git a/core/network/space.go b/core/network/space.go
index 291b431fd4..3010d6f079 100644
--- a/core/network/space.go
+++ b/core/network/space.go
@@ -50,6 +50,13 @@ type SpaceInfo struct {
 // SpaceInfos is a collection of spaces.
 type SpaceInfos []SpaceInfo
 
+// AllSpaceInfos satisfies the SpaceLookup interface.
+// It is useful for passing to conversions where we already have the spaces
+// materialised and don't need to pull them from the DB again.
+func (s SpaceInfos) AllSpaceInfos() (SpaceInfos, error) {
+	return s, nil
+}
+
 // String returns returns a quoted, comma-delimited names of the spaces in the
 // collection, or <none> if the collection is empty.
 func (s SpaceInfos) String() string {
```

Reply:
Thanks for the tip. I have updated the code to use this pattern and dropped the modifications to the `ProviderAddress`.